src/basic/smack-util.c

```diff
@@ -45,6 +45,75 @@ bool mac_smack_use(void) {
 #endif
 }
 
+static int mac_get_xattr(const char *path, const char *attr, const char **label) {
+        char *l = NULL;
+        ssize_t s;
+
+        assert(path);
+        assert(attr);
+        assert(label);
+
+        l = new0(char, NAME_MAX);
+        if (!l)
+                return -ENOMEM;
+
+        s = getxattr(path, attr, l, NAME_MAX);
+        if (s < 0) {
+                free(l);
+                return -errno;
+        }
+
+        l = realloc(l, (size_t) s + 1);
+        if (!l)
+                return -ENOMEM;
+
+        *label = l;
+
+        return s;
+}
+
+static int mac_get_xattr_fd(int fd, const char *attr, const char **label) {
+        char *l = NULL;
+        ssize_t s;
+
+        assert(fd >= 0);
+        assert(attr);
+        assert(label);
+
+        l = new0(char, NAME_MAX);
+        if (!l)
+                return -ENOMEM;
+
+        s = fgetxattr(fd, attr, l, NAME_MAX);
+        if (s < 0) {
+                free(l);
+                return -errno;
+        }
+
+        l = realloc(l, (size_t) s);
+        if (!l)
+                return -ENOMEM;
+
+        *label = l;
+
+        return s;
+}
+
+int mac_smack_read(const char *path, const char **label) {
+        int r = 0;
+
+#ifdef HAVE_SMACK
+        if (!mac_smack_use())
+                return 0;
+
+        r = mac_get_xattr(path, "security.SMACK64", label);
+        if (r < 0)
+                return r;
+#endif
```

Comment:
This pattern repeats multiple times. You can move it to its own function and stub the entire thing out for `!HAVE_SMACK`. That way, the `#ifdef` is at once central place.